src/Middleware/HttpLogging/src/W3CLogger.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Options;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using System.Numerics;
+using System.Text;
+using Microsoft.Net.Http.Headers;
+using System.Globalization;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal class W3CLogger : IAsyncDisposable
+    {
+        private readonly W3CLoggerProcessor _messageQueue;
+        private readonly IOptionsMonitor<W3CLoggerOptions> _options;
+
+        public W3CLogger(IOptionsMonitor<W3CLoggerOptions> options, IHostEnvironment environment, ILoggerFactory factory)
+        {
+            _options = options;
+            _messageQueue = InitializeMessageQueue(_options, environment, factory);
+        }
+
+        // Virtual for testing
+        internal virtual W3CLoggerProcessor InitializeMessageQueue(IOptionsMonitor<W3CLoggerOptions> options, IHostEnvironment environment, ILoggerFactory factory)
+        {
+            return new W3CLoggerProcessor(options, environment, factory);
+        }
+
+        public async ValueTask DisposeAsync() => await _messageQueue.DisposeAsync();
+
+        public void Log(string[] elements)
+        {
+            _messageQueue.EnqueueMessage(Format(elements));
+        }
+
+        private string Format(string[] elements)
+        {
+            // Need to calculate TimeTaken now, if applicable
+            var date = elements[W3CLoggingMiddleware._dateIndex];
+            var time = elements[W3CLoggingMiddleware._timeIndex];
+            if (!string.IsNullOrEmpty(date) && !string.IsNullOrEmpty(time) && _options.CurrentValue.LoggingFields.HasFlag(W3CLoggingFields.TimeTaken))
+            {
+                DateTime start = DateTime.ParseExact(date + time, "yyyy-MM-ddHH:mm:ss", CultureInfo.InvariantCulture);
+                var elapsed = DateTime.Now.Subtract(start);
+                elements[W3CLoggingMiddleware._timeTakenIndex] = elapsed.TotalMilliseconds.ToString(CultureInfo.InvariantCulture);
+            }
+
+            // 200 is around the length of an average cookie-less entry
+            var sb = new ValueStringBuilder(200);
+            var firstElement = true;
+            for (var i = 0; i < elements.Length; i++)
+            {
+                if (_options.CurrentValue.LoggingFields.HasFlag((W3CLoggingFields)(1 << i)))
```

Comment:
We should cache the logging fields (and the rest of the options) until we're notified of a change (unless we decide not to do that). Each reference to `_options.CurrentValue` is doing a ConcurrentDictionary lookup and this is accessed a lot per request.

Reply:
Done